articles/055_ros_parameter_design.md

```diff
@@ -0,0 +1,200 @@
+
+### Introspection of available parameters
+
+Provide a list of currently set parameters.
+
+### Notifications of change
+
+When a parameter changes value it should be possible to get a notification.
+
+### Reject parameter changes
+
+If a parameter change is requested to an invalid value, the parameter server should be able to reject the change.
+Related to this, it should also be able to communicate the preconditions which are necessary but not sufficient to determine the acceptance of a change.
+
+### Introspection of expected parameters
+
+When using parameters, providing a list of expected parameters can prevent accidentally setting parameters with typos in their names etc.
+
+### Control parameter lifetime
+
+When parameters are set, there should be an understanding of what the lifetime of the parameter value will be and what conditions will clear its value.
+
+### Unambigious naming
+
+All parameters should be unambiguously addressable.
+
+For example, one of the challenges of the current system is that there is a naming ambiguity between nodes and parameters `/foo/bar/baz` could be a node `/foo/bar/baz` or a private parameter `baz` on node `/foo/bar`.
+
+### Logging
+
+All changes to parameters should be visible when logging.
+And when playing back a log file the parameter changes should be able to be reapplied.
+
+## Proposed Approach
+
+To cover the feature set above, the ROS 2.0 parameter system is proposed as follows.
+
+### Parameters Hosted on nodes
+
+For the sake of validating parameter lifecycle, all parameters will be hosted on a node.
+Their lifetime will be implicitly tied to the nodes lifetime.
+The node will be responsible for validating current values.
+
+### Parameter addressing
+
+All parameters will be addressed by two elements: the full node name and the parameter name.
+
+### Supported datatypes
+
+Each parameter consists of a key and a value.
+The key is a string value.
+The value can be one of the following datatypes:
+
+- `float64`
+- `int64`
+- `string`
+- `bool`
+- `bytes[]`
+
+The datatypes are chosen as non-complex primatives.
+The full complement of datatypes of different bitdepth and unsigned types are avoided to allow interpretation from text based configuration files.
+
+`bytes` are included to allow the storage of binary blobs.
+It's use is not recommended but can be very convenient, and explicitly supporting it is better than having people try to abuse other datatypes such as strings.
+
+### Core API
+
+Each node is responsible for providing the following functionality.
+
+- **Get Parameters**
+  Given a list of parameter names it will return the parameter values.
+- **Set Parameters**
+  Given a list of parameter names, it will request an update of the values subject to validation of the values.
+  The updated values can include unsetting the value.
+  It will provide an API that can atomically update a set of values such that if any of the values fail validation, none of the values are set.
```

Comment:
I think it would be clearer to have `get`, `get atomic`, `set`, and `set atomic` all called out separately, rather than making notes about atomic versions in the description.